source/backtraceMetric.ts

```diff
@@ -0,0 +1,226 @@
+import { BacktraceClientOptions } from '.';
+import { uuid } from './utils';
+
+declare const __VERSION__: string;
+
+/**
+ * Function to create instance of BacktraceMetric and ping.
+ */
+export default function backtraceMetric(
+  configuration: BacktraceClientOptions,
+): void {
+  new BacktraceMetric(configuration).ping();
+}
+
+/**
+ * Handles Backtrace Metrics.
+ * TODO: Consider converting to a static class since an instance is only created to call a single function once.
+ */
+export class BacktraceMetric {
+  public universe;
+  public token;
+  public timeout = 15000; // fifteen seconds in miliseconds
+
+  /** Seconds since epoch. */
+  public readonly CURRENT_TIMESTAMP: number = Math.floor(
+    new Date().getTime() / 1000,
+  );
+  /** Thirty minutes in seconds. */
+  public readonly THIRTY_MINUTES = 1800;
+
+  public readonly userAgent = navigator.userAgent;
+  public readonly applicationName = 'backtrace-js';
+  public readonly applicationVersion = __VERSION__;
+
+  // These fields are stored in localStorage for persistence.
+  public sessionId = this.getSessionId();
+  public lastActive = this.getLastActive();
+  public sessionStart = this.getSessionStart();
+
+  private DIRECTORY = 'events-test'; // api endpoint ex. 'https://<directory>.backtrace.io/api/...'. Should just be 'events'
+
+  constructor(configuration: BacktraceClientOptions) {
+    const universe = ''; // get universe name from configuration if possible. If not, pass in / access universe name somewhere.
+
+    if (!configuration.endpoint) {
+      throw new Error(`Backtrace: missing 'endpoint' option.`);
+    }
+    if (!configuration.token) {
+      throw new Error(`Backtrace: missing 'token' option.`);
+    }
+
+    this.universe = universe;
+    this.token = configuration.token;
+    this.timeout = configuration.timeout;
+  }
+
+  /**
+   * Handle sessions and events. When pinged / called, will create or manage current session and send unique/summed events
+   * when appropriate.
+   */
+  public ping() {
+    this.setLastActive(this.CURRENT_TIMESTAMP); // update lastActive since ping was called, user is active
+
+    // If sessionId is not set, create new session. Send unique and app launch events.
+    if (!this.sessionId) {
+      const newSessionId = this.createNewSession();
+      this.sendUniqueEvent(newSessionId);
+      // An "application launch" loosely / temporarily means first session creation.
+      this.sendSummedEvent(newSessionId, 'application-launches');
+
+      // If sessionId is set and lastActive is over 30 minutes ago, create new session only.
+    } else if (this.lastActive < this.CURRENT_TIMESTAMP - this.THIRTY_MINUTES) {
```

Comment:
Based on my comment above I think we should check this if statement.

Reply:
I have refactored. I noticed I was treated "session id" as "unique user id".

With the addition of a guid for the user, and a sessionId to track sessions, I think the logic is improved now.
I also added a heartbeat so we do not rely on page navigations to persist the sessions.